zcash_client_backend/src/data_api/chain.rs

```diff
@@ -324,58 +315,77 @@ where
+    // Get the nullifiers for the unspent notes we are tracking
+    let mut nullifiers = Nullifiers::new(
+        data_db
+            .get_sapling_nullifiers(NullifierQuery::Unspent)
+            .map_err(Error::Wallet)?,
+        #[cfg(feature = "orchard")]
+        data_db
+            .get_orchard_nullifiers(NullifierQuery::Unspent)
+            .map_err(Error::Wallet)?,
+    );
 
     let mut scanned_blocks = vec![];
-    let mut scan_end_height = from_height;
-    let mut received_note_count = 0;
-    let mut spent_note_count = 0;
+    let mut scan_summary = ScanSummary::for_range(from_height..from_height);
     block_source.with_blocks::<_, DbT::Error>(
         Some(from_height),
         Some(limit),
         |block: CompactBlock| {
-            scan_end_height = block.height() + 1;
-            let scanned_block = scan_block_with_runner(
+            scan_summary.scanned_range.end = block.height() + 1;
+            let scanned_block = scan_block_with_runners::<_, _, (), ()>(
                 params,
                 block,
-                &ivks,
-                &sapling_nullifiers,
+                &scanning_keys,
+                &nullifiers,
                 prior_block_metadata.as_ref(),
-                Some(&mut batch_runner),
+                Some(&mut runners),
             )
             .map_err(Error::Scan)?;
 
-            let (s, r) = scanned_block
+            for wtx in &scanned_block.transactions {
+                scan_summary.spent_sapling_note_count += wtx.sapling_spends().len();
+                scan_summary.received_sapling_note_count += wtx.sapling_outputs().len();
+                #[cfg(feature = "orchard")]
+                {
+                    scan_summary.spent_orchard_note_count += wtx.orchard_spends().len();
+                    scan_summary.received_orchard_note_count += wtx.orchard_outputs().len();
+                }
+            }
+
+            let sapling_spent_nf: Vec<&sapling::Nullifier> = scanned_block
                 .transactions
                 .iter()
-                .fold((0, 0), |(s, r), wtx| {
-                    (s + wtx.sapling_spends.len(), r + wtx.sapling_outputs.len())
-                });
-            spent_note_count += s;
-            received_note_count += r;
-
-            let spent_nf: Vec<&sapling::Nullifier> = scanned_block
-                .transactions
-                .iter()
-                .flat_map(|tx| tx.sapling_spends.iter().map(|spend| spend.nf()))
+                .flat_map(|tx| tx.sapling_spends().iter().map(|spend| spend.nf()))
                 .collect();
+            nullifiers.retain_sapling(|(_, nf)| !sapling_spent_nf.contains(&nf));
+            nullifiers.extend_sapling(scanned_block.transactions.iter().flat_map(|tx| {
+                tx.sapling_outputs().iter().flat_map(|out| {
+                    out.key_source()
+                        .account()
+                        .zip(out.nf().copied())
+                        .into_iter()
```

Comment:
This appears to lose information. With the new `KeySource` enum, we can now be scanning with an imported FVK, so `out.nf().is_some()` but `out.key_source.account().is_none()`. In this case, the output of this will be `None`, dropping the actually-known nullifier.

I do not yet know how the new `Nullifiers` type works, but it seems to me that if we want to support imported FVKs, `Nullifiers::extend_sapling` needs to take `(Option<AccountId>, Nullifier)` instead of `(AccountId, Nullifier)`. Alternatively, if the `Nullifiers` type is as intended, then we need to ensure we fail here if an imported FVK is used:
```suggestion
                    let account = out.key_source().account().ok_or(SomeError)?;
                    out.nf()
                        .copied()
                        .map(|nf| (account, nf))
                        .into_iter()
```

Reply:
Digging into this a bit, the sole reason that `WalletSpend` has an `account` field is so that we can use the account id in the `is_change` check; we consider a value change when the account that received a note also spent notes in the same transaction. For the FVK use case, ideally this would be a unique identifier for the FVK, rather than the account ID, because we want to be able to detect change reliably irrespective of whether or not the FVK was derived from the wallet's seed.

Essentially, I think what this should actually be is the key fingerprint, not something derived from the key source at all.

Reply:
I continue to be frustrated at how the scope of the next `zcash_client_backend` release keeps being expanded. I would much prefer that we just do not support the imported FVK use case for this release, and then make this change in the next one.

Reply:
So here's the thing: that extraction, in https://github.com/zcash/librustzcash/pull/1202, is relatively trivial right now, because since this PR evolves incrementally from the existing API each step of a rebase of this PR atop #1202 is tractable. The alternative, of rewriting #1202 and then #1175 atop this PR, is substantially more complex. I regret not having done that extraction before I started this work, but I think that in the interest of saving the labor it might be worth doing now.

I guess that the alternative, somewhat weirder approach is to return the error here, but _also_ do the rebase, and then after this PR merges and we do the release, then I can reconstruct just the delta by squash-merging the rebased branch to a new branch from `main`. Which... I guess is feasible, so long as nothing _else_ conflicting gets merged in between.